edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Diagnostics/publisher/EdgeRuntimeDiagnosticsUpload.cs

```diff
@@ -25,26 +28,43 @@ public EdgeRuntimeDiagnosticsUpload(IEdgeAgentConnection edgeAgentConnection)
         public async Task<bool> PublishAsync(IEnumerable<Metric> metrics, CancellationToken cancellationToken)
         {
             Preconditions.CheckNotNull(metrics, nameof(metrics));
-            byte[] data = MetricsSerializer.MetricsToBytes(metrics).ToArray();
 
-            if (data.Length > 0)
+            Message[] messagesToSend = this.BatchAndBuildMessages(metrics.ToArray()).ToArray();
+            try
             {
-                Message message = this.BuildMessage(data);
-
-                try
-                {
-                    await this.edgeAgentConnection.SendEventAsync(message);
-                }
-                catch (Exception ex) when (ex.HasTimeoutException())
-                {
-                    Log.LogDebug(ex, "Send message to IoTHub");
-                    return false;
-                }
+                await this.edgeAgentConnection.SendEventBatchAsync(messagesToSend);
+            }
+            catch (Exception ex) when (ex.HasTimeoutException())
+            {
+                Log.LogDebug(ex, "Send message to IoTHub");
+                return false;
             }
 
             return true;
         }
 
+        IEnumerable<Message> BatchAndBuildMessages(ArraySegment<Metric> metrics)
+        {
+            if (!metrics.Any())
+            {
+                return Enumerable.Empty<Message>();
+            }
+
+            byte[] data = MetricsSerializer.MetricsToBytes(metrics).ToArray();
+
+            if (data.Length > MaxMessageSize)
+            {
+                var part1 = this.BatchAndBuildMessages(metrics.Slice(0, metrics.Count / 2));
+                var part2 = this.BatchAndBuildMessages(metrics.Slice(metrics.Count / 2, metrics.Count / 2 + metrics.Count % 2));
```

Comment:
There's another overload of ArraySegment.Slice that only takes an index arg, and returns a slice from the index to the end of the parent slice. Seems like a better option than doing the math yourself for `part2`.